src/dynamics/mpas/dyn_comp.F90

```diff
@@ -139,13 +174,640 @@ pure function sec_to_hour_min_sec(sec) result(hour_min_sec)
+            else
+               call endrun('Failed to find variable "PHIS"', subname, __LINE__)
+            end if
+
+            ! Surface geometric height in MPAS should match the values in topography file.
+            if (any(abs(zgrid(1, 1:ncells_solve) - surface_geometric_height) > error_tolerance)) then
+                call endrun('Surface geometric height in MPAS is not consistent with topography data', subname, __LINE__)
+            end if
+
+            deallocate(surface_geopotential)
+            deallocate(surface_geometric_height)
+        else
+            call dyn_debug_print('Topography file is not used')
+
+            ! Surface geometric height in MPAS should be zero.
+            if (any(abs(zgrid(1, 1:ncells_solve)) > error_tolerance)) then
+                call endrun('Surface geometric height in MPAS is not zero', subname, __LINE__)
+            end if
+        end if
+
+        nullify(zgrid)
+    end subroutine check_topography_data
+
+    !> Set analytic initial condition for MPAS.
+    !> (KCW, 2024-05-22)
+    subroutine set_analytic_initial_condition()
+        character(*), parameter :: subname = 'dyn_comp::set_analytic_initial_condition'
+        integer, allocatable :: global_grid_index(:)
+        real(kind_r8), allocatable :: buffer_2d_real(:, :), buffer_3d_real(:, :, :)
+        real(kind_r8), allocatable :: lat_rad(:), lon_rad(:)
+        real(kind_r8), allocatable :: z_int(:, :)       ! Geometric height (meters) at layer interfaces.
+                                                        ! Dimension and vertical index orders follow CAM-SIMA convention.
+        real(kind_r8), pointer :: zgrid(:, :) => null() ! Geometric height (meters) at layer interfaces.
+                                                        ! Dimension and vertical index orders follow MPAS convention.
+
+        call init_shared_variable()
+
+        call set_mpas_state_u()
+        call set_mpas_state_w()
+        call set_mpas_state_scalars()
+        call set_mpas_state_rho_theta()
+        call set_mpas_state_rho_base_theta_base()
+
+        deallocate(global_grid_index)
+        deallocate(lat_rad)
+        deallocate(lon_rad)
+        deallocate(z_int)
+
+        nullify(zgrid)
+    contains
+        !> Initialize variables that are shared and repeatedly used by the `set_mpas_state_*` internal subroutines.
+        !> (KCW, 2024-05-13)
+        subroutine init_shared_variable()
+            integer :: ierr
+            integer :: k
+            integer, pointer :: indextocellid(:) => null()
+            real(kind_r8), pointer :: lat_deg(:) => null(), lon_deg(:) => null()
+
+            call dyn_debug_print('Preparing to set analytic initial condition')
+
+            allocate(global_grid_index(ncells_solve), stat=ierr)
+            call check_allocate(ierr, 'set_analytic_initial_condition', 'global_grid_index(ncells_solve)', 'dyn_comp', __LINE__)
```

Comment:
Would it be worth defining a `subname` parameter for `'set_analytic_conditions::init_shared_variables'` to use in calls to `check_allocate` throughout this routine?

Reply:
I pushed https://github.com/ESCOMP/CAM-SIMA/pull/267/commits/d18264ba2467694d053246789215e23c2048f176 to also add `subname` to internal subroutines.

Note that with this commit, `fortls` will complain about variable shadowing ("Variable "subname" masks variable in parent scope"). It should be fine because that is just a diagnostic.